### PR TITLE
Support using disabled data source at startup(#20738)

### LINKED ISSUE #20738: Support using disabled data source at startup
Now we can use DistSQL to disable the datasource's state in read-write splitting rules.
For example,proxy A has disabled the data source.Then proxy B is started with the same configuration, and the status of the disabled data source will still be checked, resulting in startup failure.

In order to ensure that the proxy can startup normally, we need to do the following tasks:

- [ ] disabled data sources will not be checked when startup
- [ ] use `--force` program parameters to ensure that the proxy starts normally regardless of whether it can be connected or not

### PR DESCRIPTION
**Pull request by FlyingZC**
Fixes #20738.

Changes proposed in this pull request:
- Support using disabled data source at startup

### CODE REVIEW

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceStateManager.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+                if (DataSourceState.DISABLED != dataSourceStates.get(getCacheKey(databaseName, key))) {
+                    result.put(key, value);
+                }
+            });
+        }
+        return result;
+    }
+    
+    private String getCacheKey(final String databaseName, final String key) {
+        return databaseName + "." + key;
```
Comment (tuichenchuxin) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceStateManager.java:149]: key name

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceState.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+     * @param state data source state name
+     * @return data source state
+     */
+    public static DataSourceState getDataSourceState(final String state) {
+        for (DataSourceState each : values()) {
+            if (each.name().equalsIgnoreCase(state)) {
+                return each;
+            }
+        }
+        return null;
```
Comment (tuichenchuxin) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceState.java:59]: exception will be better?

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceStateManager.java (hunk 1 above)
Comment (FlyingZC) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceStateManager.java:149]: Your are right.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceState.java (hunk 2 above)
Comment (FlyingZC) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceState.java:59]: Your are right.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/type/DatabaseTypeEngine.java [hunk 3]
```diff
@@ -44,12 +44,14 @@ public final class DatabaseTypeEngine {
...
+     *
+     * @param databaseName database name
      * @param databaseConfig database configuration
      * @param props props
      * @return protocol type
      */
-    public static DatabaseType getProtocolType(final DatabaseConfiguration databaseConfig, final ConfigurationProperties props) {
-        return findConfiguredDatabaseType(props).orElseGet(() -> getDatabaseType(databaseConfig.getDataSources().values()));
+    public static DatabaseType getProtocolType(final String databaseName, final DatabaseConfiguration databaseConfig, final ConfigurationProperties props) {
+        return findConfiguredDatabaseType(props).orElseGet(() -> getDatabaseType(DataSourceStateManager.getInstance().getNonDisabledDataSources(databaseName, databaseConfig)));
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/type/DatabaseTypeEngine.java:54]: Do you think `getEnabledDataSources` is better?

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/instance/metadata/jdbc/JDBCInstanceMetaData.java [hunk 4]
```diff
@@ -32,9 +32,18 @@ public final class JDBCInstanceMetaData implements InstanceMetaData {
     
     private final String ip;
     
+    private final boolean force;
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/instance/metadata/jdbc/JDBCInstanceMetaData.java:35]: Why add force in JDBCInstanceMetaData?

File: shardingsphere-proxy/shardingsphere-proxy-bootstrap/src/main/java/org/apache/shardingsphere/proxy/arguments/BootstrapArguments.java [hunk 5]
```diff
@@ -74,6 +74,19 @@ public List<String> getAddresses() {
...
+     * Get force startup parameter.
+     *
+     * @return force parameter
+     */
+    public boolean getForce() {
+        return args.length >= 4 && parseForceParameter(args[3]);
+    }
+    
+    private boolean parseForceParameter(final String forceParam) {
+        return forceParam.trim().equals("-f");
```
Comment (strongduanmu) [inline shardingsphere-proxy/shardingsphere-proxy-bootstrap/src/main/java/org/apache/shardingsphere/proxy/arguments/BootstrapArguments.java:87]: Can we start up Proxy with start.sh?

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/type/DatabaseTypeEngine.java (hunk 3 above)
Comment (FlyingZC) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/type/DatabaseTypeEngine.java:54]: You are right

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/instance/metadata/jdbc/JDBCInstanceMetaData.java (hunk 4 above)
Comment (FlyingZC) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/instance/metadata/jdbc/JDBCInstanceMetaData.java:35]: You are right.

File: shardingsphere-proxy/shardingsphere-proxy-bootstrap/src/main/java/org/apache/shardingsphere/proxy/arguments/BootstrapArguments.java (hunk 5 above)
Comment (FlyingZC) [reply shardingsphere-proxy/shardingsphere-proxy-bootstrap/src/main/java/org/apache/shardingsphere/proxy/arguments/BootstrapArguments.java:87]: I will try.

**Comment by FlyingZC** [comment]
image

**Review by tuichenchuxin — APPROVED** [review]

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/type/DatabaseTypeEngine.java [hunk 6]
```diff
@@ -44,12 +44,14 @@ public final class DatabaseTypeEngine {
     /**
      * Get protocol type.
      * 
+     *
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/type/DatabaseTypeEngine.java:47]: Please remove this useless line.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/type/DatabaseTypeEngine.java [hunk 7]
```diff
@@ -64,9 +66,7 @@ public static DatabaseType getProtocolType(final Map<String, ? extends DatabaseC
         if (configuredDatabaseType.isPresent()) {
             return configuredDatabaseType.get();
         }
-        Collection<DataSource> dataSources = databaseConfigs.values().stream()
-                .filter(each -> !each.getDataSources().isEmpty()).findFirst().map(optional -> optional.getDataSources().values()).orElseGet(Collections::emptyList);
-        return getDatabaseType(dataSources);
+        return getDatabaseType(DataSourceStateManager.getInstance().getEnabledDataSources(databaseConfigs));
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/type/DatabaseTypeEngine.java:69]: Why is the databaseName parameter not needed here?

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceState.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    DISABLED, ENABLED;
+    
+    /**
+     * Data source disable or enable.
+     *
+     * @param status data source state
+     * @return disable or enable
+     */
+    public static boolean isDisable(final String status) {
+        return DISABLED.name().toLowerCase().equals(status);
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceState.java:34]: Do  you think equalsIgnoreCase is better?

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceState.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    }
+    
+    /**
+     * Data source disable or enable.
+     *
+     * @param status storage node status
+     * @return disable or enable
+     */
+    public static boolean isEnable(final String status) {
+        return ENABLED.name().toLowerCase().equals(status);
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceState.java:44]: Do  you think equalsIgnoreCase is better?

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceState.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    }
+    
+    /**
+     * Get data source state by state name.
+     *
+     * @param state data source state name
+     * @return data source state
+     */
+    public static DataSourceState getDataSourceState(final String state) {
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceState.java:53]: Is it possible to use map to improve performance?

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceStateManager.java [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+     * Set data source states when bootstrap.
+     *
+     * @param databaseName database name
+     * @param dataSources data sources
+     * @param storageDataSourceStates storage node data source status
+     * @param force whether to force start
+     */
+    public void initStates(final String databaseName, final Map<String, DataSource> dataSources, final Map<String, DataSourceState> storageDataSourceStates, final boolean force) {
+        this.force = force;
+        dataSources.forEach((key, value) -> {
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceStateManager.java:70]: Is it possible to reduce the hierarchy by splitting the method?

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceStateManager.java [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+        initialized = true;
+    }
+    
+    /**
+     * Get enabled data sources.
+     *
+     * @param databaseConfigs database config
+     * @return enabled data sources
+     */
+    public Collection<DataSource> getEnabledDataSources(final Map<String, ? extends DatabaseConfiguration> databaseConfigs) {
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceStateManager.java:95]: databaseConfigs represents multiple logical libraries, is there a problem with the processing logic here?

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/instance/metadata/jdbc/JDBCInstanceMetaData.java (hunk 4 above)
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/instance/metadata/jdbc/JDBCInstanceMetaData.java:35]: @zhaojinchao95 Can you help review this change?

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/instance/metadata/proxy/ProxyInstanceMetaData.java [hunk 13]
```diff
@@ -37,17 +37,28 @@ public final class ProxyInstanceMetaData implements InstanceMetaData {
     
     private final int port;
     
+    private final boolean force;
+
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/instance/metadata/proxy/ProxyInstanceMetaData.java:41]: @zhaojinchao95 Can you help review this change?

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/database/resource/ShardingSphereResource.java [hunk 14]
```diff
@@ -39,16 +40,19 @@
 @Getter
 public final class ShardingSphereResource {
     
+    private final String databaseName;
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/database/resource/ShardingSphereResource.java:43]: Why add databaseName here?

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/type/DatabaseTypeEngine.java (hunk 6 above)
Comment (FlyingZC) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/type/DatabaseTypeEngine.java:47]: Ok.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceState.java (hunk 8 above)
Comment (FlyingZC) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceState.java:34]: Ok.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceState.java (hunk 9 above)
Comment (FlyingZC) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceState.java:44]: Ok.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceState.java (hunk 10 above)
Comment (FlyingZC) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceState.java:53]: Ok.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/instance/metadata/jdbc/JDBCInstanceMetaData.java (hunk 4 above)
Comment (zhaojinchao95) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/instance/metadata/jdbc/JDBCInstanceMetaData.java:35]: I think the `force` parameter move to props is better.

**Review by zhaojinchao95 — APPROVED** [review]

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/type/DatabaseTypeEngine.java (hunk 7 above)
Comment (FlyingZC) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/type/DatabaseTypeEngine.java:69]: You are right.I shouldn't manage logic other than data source state.I will change the api.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceStateManager.java (hunk 12 above)
Comment (FlyingZC) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceStateManager.java:95]: The unique data source state key is `databaseName + dataSourceName`.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/database/resource/ShardingSphereResource.java (hunk 14 above)
Comment (FlyingZC) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/database/resource/ShardingSphereResource.java:43]: I will remove it.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceStateManager.java (hunk 11 above)
Comment (FlyingZC) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceStateManager.java:69]: You are right.

**Review by tuichenchuxin — APPROVED** [review]

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceStateManager.java [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+     * @return enabled data source map
+     */
+    public Map<String, DataSource> getEnabledDataSourceMap(final String databaseName, final Map<String, DataSource> dataSources) {
+        if (dataSources.isEmpty() || !initialized) {
+            return dataSources;
+        }
+        Map<String, DataSource> result = new LinkedHashMap<>(dataSources.size(), 1);
+        dataSources.forEach((key, value) -> {
+            DataSourceState dataSourceState = dataSourceStates.get(getCacheKey(databaseName, key));
+            if (force ? DataSourceState.ENABLED == dataSourceState : DataSourceState.DISABLED != dataSourceState) {
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceStateManager.java:121]: Please extract this logic to a param out of foreach loop.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceStateManager.java (hunk 15 above)
Comment (FlyingZC) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceStateManager.java:121]: Ok

**Review by strongduanmu — APPROVED** [review]
LGTM.

**Review by strongduanmu — APPROVED** [review]

**Pull request merged**